components/core/src/clp/StreamingReader.hpp

```diff
@@ -0,0 +1,339 @@
+        }
+    };
+
+    /**
+     * Enum defining the states of the reader.
+     */
+    enum class State : uint8_t {
+        // InProgress: The reader is transferring data from the source URL.
+        InProgress = 0,
+        // Failed: The reader has failed to stream data from the source URL.
+        Failed,
+        // Finished: The streaming of the given source URL has been accomplished.
+        Finished
+    };
+
+    static constexpr size_t cDefaultBufferPoolSize{8};
+    static constexpr size_t cDefaultBufferSize{4096};
+
+    static constexpr size_t cMinBufferPoolSize{2};
+    static constexpr size_t cMinBufferSize{512};
+
+    /**
+     * These static members defines the default connection timeout and operation timeout.
+     * Please refer to the libcurl documentation for more details:
+     * ConnectionTimeout: https://curl.se/libcurl/c/CURLOPT_CONNECTTIMEOUT.html
+     * OverallTimeout: https://curl.se/libcurl/c/CURLOPT_TIMEOUT.html
+     */
+    static constexpr uint32_t cDefaultConnectionTimeout{10};  // Seconds
+    static constexpr uint32_t cDefaultOverallTimeout{45};  // Seconds
+
+    /**
+     * Initializes the underlying libcurl functionalities globally. It should be called before
+     * opening a source URL for data transferring.
+     * @return ErrorCode_Success on success.
+     * @return ErrorCode_Failure if libcurl initialization failed.
+     */
+    [[nodiscard]] static auto init() -> ErrorCode;
+
+    /**
+     * Releases the globally initialized libcurl resources.
+     */
+    static auto deinit() -> void;
+
+    /**
+     * Constructs a new connection to stream data from the given url with a given offset.
+     * TODO: the current implementation doesn't handle the case when the given offset is out of
+     * range. The file_pos will be set to an invalid state if this happens, which can be
+     * problematic if the other part of the program depends on this position. It can be fixed by
+     * capturing the error code 416 in the response header.
+     * @param src_url
+     * @param offset The offset of the starting byte used for data transferring.
+     * @param disable_caching Whether to disable the caching.
+     * @param overall_timeout_in_sec The overall timeout in seconds used by the underlying CURL
+     * handler. Doc: https://curl.se/libcurl/c/CURLOPT_TIMEOUT.html
+     * @param connection_timeout_in_sec The connection timeout in seconds used by the underlying
+     * CURL handler. Doc: https://curl.se/libcurl/c/CURLOPT_CONNECTTIMEOUT.html
+     * @param buffer_pool_size Total number of buffers available for transferring.
+     * @param buffer_size The size of each data buffer.
+     */
+    explicit StreamingReader(
+            std::string_view src_url,
+            size_t offset = 0,
+            bool disable_caching = false,
+            uint32_t overall_timeout_int_sec = cDefaultOverallTimeout,
+            uint32_t connection_timeout_in_sec = cDefaultConnectionTimeout,
```

Comment:
we can make them consistent, but I prefer to leave `offset` and `disable_caching` in the front since users have a higher chance to not use default values for these 2 options

Reply:
also `CurlDownloaderHandler` is now a generic class. Making both have the same ordering doesn't really matter IMO. If we want them to be 100% match we shouldn't make them two classes

Reply:
Reordered as suggested except prioritizing `offset` and `disable_caching`